### PR TITLE
feat: add session support

### LINKED ISSUE #6375: "Removing cookie failed for" error in Firefox after setting cookie via `cy.setCookie()`
### Current behavior:
One in every ten or so runs on Firefox I get the following error:

image

### Test code to reproduce:
In our `support/index.js` file we have a `beforeEach` that uses `cy.setCookie()` to set a cookie that opts our automation out of any experiments that could be running on our site. Not sure if this is actually the reason for it since we are adding a cookie but the error says removing cookie, but that is the only place that has our main hostname instead of our test host

### Versions
Cypress 4.0.0
Firefox 72.0.2

### LINKED ISSUE #8301: Add `session` API for snapshotting/restoring session related data (cookies, localStorage, sessionStorage)
## Summary

We would like to implement solutions for authentication and switching between user sessions.

To accomplish those objectives, we need to save checkpoints that represent the browser’s state at the end of some steps the user has performed. After saving the state we will be able to go back and forth between points in time.

For the time being we will call this a **Session**, though the name is up for debate.

---

Currently, if your tests rely on having set session related data (e.g. cookies, localStorage, sessionStorage), you have two choices:

- "build up" session data through UI interactions (similar to Page Object Pattern). For multiple tests that use the same session data, this leads to either slower tests (steps done in `beforeEach` every test) or leads to tests that can't run in isolation (requiring previous tests to build up state).
- use `cy.request`/`cy.setCookie` to programmatically set session data. For multiple tests, this reduces execution time drastically compared to Page Object Pattern. However this requires more knowledge about your app's auth to setup, can potentially be a maintenance burden, and covers less your app.

`cy.session` is meant to merge the simplicity of Page-Object-Pattern to the execution speed of programmatic auth. With `cy.session`, you'll be able to perform UI interactions before taking a "snapshot" of all the session related data and "restore" it before each test. 

### Defining a Session 
Sessions should be defined at the root level of your Cypress project, within `cypress/sessions.js`. A best practice is to define all of your sessions here instead of in individual spec files.

### Defining a Single Session

The proposed API for defining a single Session looks like this:

```ts
export const admin = cy.defineSession({
  name: 'admin',
  steps() {
    cy.get('#username').type('my-username')
    cy.get('#password').type('my-password')
    cy.get('#login').click()
  },
})
```

### Defining Multiple Sessions

Some alternate syntaxes utilize a `Cypress.defineSessions()` method to add many sessions at once. The keys in this object are the session names, and the value can either be a function or a Session Object.

```ts
// The keys in this map to session names
// The values may either be Session object or a Steps function
export const sessions = cy.defineSessions({
  admin() {
    cy.get('#username').type('admin-username')
    cy.get('#password').type('admin-password')
    cy.get('#login').click()
  },
  regularUser: {
    steps() {
      cy.get('#username').type('basic-username')
      cy.get('#password').type('basic-password')
      cy.get('#login').click()
    }
  }
}
```

### Assuming a Session

Assuming a session will be done in either a `beforeEach` or within a block (`it`/`describe`), so that each test starts from a single session, or programmatically whenever a user wants.

There are two ways to assume a session. One takes advantage of per-test configuration and the other is a method on the `cy` object. The API design goal is for sessions to be real Type-able objects that can be used. We will still support string-lookup of sessions by name, but using real JavaScript objects is preferred.

### Programmatic API

By reference

```ts
import { admin } from './sessions'

// Switch between sessions programmatically using the Session's object
cy.useSession(admin)

// Remove previous state, rebuild and then use the session
cy.useSession(admin, { fresh: true })
```

By string
```ts
// Switch between sessions programmatically using Session.name
cy.useSession('admin')
```
Per-test configuration
By reference

```ts
import { admin } from './sessions'
describe('As an administrator', { session: admin }, () => {
  /* ... */
})
By string
describe('As an administrator', { session: 'admin' }, () => {
  /* ... */
})
```
### Retrying Corrupt Sessions

When assuming a session throws an error, it is considered corrupt. Sessions may be corrupt for a number of reasons: Bugs in the Steps logic, corrupt state from previous tests, invalid login tokens, or general flake.

If this happens, the session’s internal state will be invalidated, and it will be rebuilt from scratch. After this, the test run will proceed as normal.

### Composition with Sessions

Because sessions are simply a set of steps and a starting state, it is possible to compose with them by reusing `cy.useSession`

To create a new user with many cart items, you’d do

```ts
const seedDb = () => cy.task('create:user')

// prefer by-function steps because adding onto
// state will possibly pollute any backends
const loginSteps = () => { /* login steps */ }
export const newUser = cy.defineSession({
  steps() {
    //  seedDb()
    /* signup */
    loginSteps
  },
  before() { seedDb() } // always exec
})

export const userWithItems = cy.defineSession({
  steps() {
    loginSteps()
    // add an item to the cart, etc
  } 
})
```

Or to create a user who opts out of tracking before signing up, you’d do: 
	
```ts
export const privacyConsciousUser = cy.defineSession({
  steps() {
    // opt out of setting cookies
    // This assumes the backend or cookies would persist this info
    // window state is NOT stored with the session.
    cy.window()
      .then((win) => win.optOut = true)

    cy.useSession(newUser)
  }
})
```

The combined state at the end of the steps function is stored as that Session’s state

### PR DESCRIPTION
**Pull request by kuceb**
- Closes #8301
- fix #6375

### User facing changelog

changes to default blank page and useSession blank page:
before | after
-|-
image | image
n/a | image

### PR Tasks

- [x]  remove user facing api changes to getCookie/setCookie, support hostOnly in backend only

- [x] Have tests been added/updated?
- [ ] Has the original issue or this PR been tagged with a release in ZenHub? 
- [x]  Docs PR: https://github.com/cypress-io/cypress-documentation/pull/3427
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [x] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/driver/src/cypress/cy.js [hunk 1]
```diff
@@ -122,6 +122,14 @@ const setTopOnError = function (Cypress, cy) {
...
 
 const commandRunningFailed = (Cypress, state, err) => {
+  // debugger
+  // return
+  // if (state('onCommandFailed')) {
+  //   state('onCommandFailed')()
+
+  //   return
+  // }
+
```
Comment (jennifer-shehane) [inline packages/driver/src/cypress/cy.js:132]: Remove commented out code or put a comment explaining why it's here.

File: packages/driver/src/cypress/cy.js [hunk 2]
```diff
@@ -659,6 +669,12 @@ const create = function (specWindow, Cypress, Cookies, state, config, log) {
       state('reject', rejectOuterAndCancelInner)
     })
     .catch((err) => {
+      // console.log('CAUGHT ERROR', err)
```
Comment (jennifer-shehane) [inline packages/driver/src/cypress/cy.js:672]: Remove commented out code or put a comment explaining why it's here.

File: packages/server/test/support/fixtures/projects/session/cypress.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+  "experimentalSessionSupport": true
```
Comment (jennifer-shehane) [inline packages/server/test/support/fixtures/projects/session/cypress.json:2]: Unnecessarily generated files - can remove them from git.

```suggestion
  "experimentalSessionSupport": true,
  "supportFile": false,
  "fixturesFolder": false
```

File: packages/driver/src/cy/commands/sessions.ts [hunk 4]
```diff
@@ -0,0 +1,822 @@
...
+            cy.then(async () => {
+              await navigateAboutBlank()
+              Cypress.log({ groupEnd: true, name: '', message: '', emitOnly: true })
+            })
+          })
+        })
+      }
+
+      const throwValidationError = (err) => {
+        err.message += '\n\nThis error occurred in a session validate hook after initializing the session. Because validation failed immediately after session setup we failed the test.'
```
Comment (chrisbreiding) [inline packages/driver/src/cy/commands/sessions.ts:752]: This should use [$errUtils.modifyErrMsg](https://github.com/cypress-io/cypress/blob/8550842e3696095c9391e89f30c87af6369ecfb9/packages/driver/src/cypress/error_utils.js#L118) so that it updates the message in the stack as well.

File: packages/driver/src/cy/commands/sessions.ts [hunk 5]
```diff
@@ -0,0 +1,822 @@
...
+            throw new Error(`unexpected option **${key}** passed to cy.session options`)
+          }
+
+          const actualType = typeof options[key]
+
+          if (actualType !== expectedType) {
+            throw new Error(`invalid option **${key}** passed to cy.session options. Expected **${expectedType}**, got ${actualType}`)
+          }
+        })
+      }
```
Comment (chrisbreiding) [inline packages/driver/src/cy/commands/sessions.ts:501]: These error messages should be in [error_messages.js](https://github.com/cypress-io/cypress/blob/8550842e3696095c9391e89f30c87af6369ecfb9/packages/driver/src/cypress/error_messages.js), utilizing `cmd('sessions')` so it's formatted consistently. They should all include the value of the argument when it's incorrect. For example, `cy.session() optional third argument must be an object, you passed: true`.

File: packages/driver/src/cy/commands/sessions.ts [hunk 6]
```diff
@@ -0,0 +1,822 @@
...
+      if (!setup) {
+        if (!existingSession || !currentTestRegisteredSessions.has(id)) {
+          $errUtils.throwErrByPath('sessions.session.not_found', { args: { id } })
+        }
+      } else {
+        const isUniqSessionDefinition = !existingSession || existingSession.setup.toString().trim() !== setup.toString().trim()
+
+        if (isUniqSessionDefinition) {
+          if (currentTestRegisteredSessions.has(id)) {
+            throw $errUtils.errByPath('sessions.session.duplicateId', { id: existingSession.id })
```
Comment (chrisbreiding) [inline packages/driver/src/cy/commands/sessions.ts:514]:
```suggestion
            $errUtils.throwErrByPath('sessions.session.duplicateId', { args: { id: existingSession.id } })
```

File: packages/driver/src/cy/commands/sessions.ts [hunk 7]
```diff
@@ -0,0 +1,822 @@
...
+            err = new Error(err)
+          }
+
+          err.stack = $stackUtils.normalizedStack(err)
+
+          err = $errUtils.enhanceStack({
+            err,
+            userInvocationStack: $stackUtils.getUserInvocationStack(err, Cypress.state),
+            projectRoot: Cypress.config('projectRoot'),
+          })
```
Comment (chrisbreiding) [inline packages/driver/src/cy/commands/sessions.ts:661]: In what cases won't an error hit `cy.fail`? It's a bit concerning to see this here since all other failures go through the [`fail` function in `cy.js`](https://github.com/cypress-io/cypress/blob/8550842e3696095c9391e89f30c87af6369ecfb9/packages/driver/src/cypress/cy.js#L812) and it does more than is duplicated here.

File: packages/driver/src/cy/commands/sessions.ts [hunk 8]
```diff
@@ -0,0 +1,822 @@
...
+  }
+
+  async function _setStorageOnOrigins (originOptions) {
+    const specWindow = cy.state('specWindow')
+
+    const current_origin = $Location.create(window.location.href).origin
+
+    const current_origin_options_index = _.findIndex(originOptions, { origin: current_origin })
+
+    if (current_origin_options_index !== -1) {
```
Comment (chrisbreiding) [inline packages/driver/src/cy/commands/sessions.ts:96]: This is a minor nitpick, but why are these snake-cased? We use camel-casing for variables everywhere else in the codebase.

File: packages/driver/src/cy/commands/sessions.ts [hunk 9]
```diff
@@ -0,0 +1,822 @@
...
+          event.source.postMessage({ type: 'set:storage:data', data: opts }, '*')
+        } else if (data.type === 'set:storage:complete') {
+          successOrigins.push(event.origin)
+          if (successOrigins.length === origins.length) {
+            resolve()
+          }
+        }
+      }
+
+      specWindow.addEventListener('message', onPostMessage)
```
Comment (chrisbreiding) [inline packages/driver/src/cy/commands/sessions.ts:163]: Some code comments would be useful here to explain why using postMessage is necessary.

File: packages/driver/src/cy/commands/local_storage.js [hunk 10]
```diff
@@ -58,5 +58,14 @@ module.exports = (Commands, Cypress, cy, state) => {
...
     },
+
+    // TODO: add these user-facing APIs
+    // getLocalStorage (options) {
+    //   return Cypress.session.getLocalStorage(options)
+    // },
+
+    // setLocalStorage (options) {
+    //   return Cypress.session.setLocalStorage(options)
+    // },
```
Comment (chrisbreiding) [inline packages/driver/src/cy/commands/local_storage.js:69]: Can we move this into an issue if it's something we'd like to add, then delete these lines from this PR? I feel like this is going to stay permanently commented out otherwise.

File: packages/reporter/src/errors/error-stack.tsx [hunk 11]
```diff
@@ -18,7 +18,7 @@ const isMessageLine = (stackLine: ParsedStackFileLine | ParsedStackMessageLine)
 }
 
 const ErrorStack = observer(({ err }: Props) => {
-  if (!err.parsedStack) return <>err.stack</>
```
Comment (chrisbreiding) [inline packages/reporter/src/errors/error-stack.tsx:21]: whoops lol 😅

File: packages/server/test/support/fixtures/projects/session/cypress/integration/spec.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+})
+
+// const login = () => {
+//   cy.visit('http://localhost:3000/login?user=user1')
+//   cy.get('input:first').type('user')
+//   cy.get('input[type=password]').type('pass')
+//   cy.get('button').click()
+
+//   cy.contains('success')
+// }
```
Comment (chrisbreiding) [inline packages/server/test/support/fixtures/projects/session/cypress/integration/spec.js:17]: Can this be deleted?

**Review by chrisbreiding — CHANGES_REQUESTED** [review]

File: packages/driver/src/cy/commands/sessions.ts [hunk 13]
```diff
@@ -0,0 +1,822 @@
...
+  const { Promise } = Cypress
+
+  const setActiveSession = (obj) => {
+    const currentSessions = cy.state('activeSessions') || {}
+
+    const newSessions = { ...currentSessions, ...obj }
+
+    cy.state('activeSessions', newSessions)
+  }
+  const getActiveSession = (id) => {
```
Comment (flotwig) [inline packages/driver/src/cy/commands/sessions.ts:60]:
```suggestion
  }
  
  const getActiveSession = (id) => {
```

prettier when?

File: packages/driver/src/cy/commands/sessions.ts [hunk 14]
```diff
@@ -0,0 +1,822 @@
...
+          successOrigins.push(event.origin)
+          if (successOrigins.length === origins.length) {
+            resolve()
+          }
+        }
+      }
+
+      specWindow.addEventListener('message', onPostMessage)
+    })
+    .timeout(2000)
```
Comment (flotwig) [inline packages/driver/src/cy/commands/sessions.ts:165]: might be worth commenting: in what error condition would this timeout ever be reached? the thread for the other origin being locked for over two seconds?

File: packages/driver/src/cy/commands/sessions.ts [hunk 15]
```diff
@@ -0,0 +1,822 @@
...
+
+    /**
+     * 1) if we only need current_origin localStorage, access sync
+     * 2) if cross-origin http, we need to load in iframe from our proxy that will intercept all http reqs at /__cypress/automation/*
+     *      and postMessage() the localStorage value to us
+     * 3) if cross-origin https, since we pass-thru https conntections in the proxy, we need to
+     *      send a message telling our proxy server to intercept the next req to the https domain,
+     *      then follow 2)
+     */
+    async getStorage (options = {}) {
```
Comment (flotwig) [inline packages/driver/src/cy/commands/sessions.ts:289]: nit: this function is >100 lines long, i get that it's a complicated procedure, but it would be easier to grok if it was broken up into smaller named functions

File: packages/driver/src/cy/commands/sessions.ts [hunk 16]
```diff
@@ -0,0 +1,822 @@
...
+          results.sessionStorage.push({ origin, value: value.sessionStorage })
+        }
+      }
+
+      const currentOriginIndex = origins.indexOf(current_origin)
+
+      if (currentOriginIndex !== -1) {
+        origins.splice(currentOriginIndex, 1)
+        // localStorage.length propery is not always accurate, we must stringify to check for entries
+        // for ex) try setting localStorge.key = 'val' and reading localStorage.length, may be 0.
```
Comment (flotwig) [inline packages/driver/src/cy/commands/sessions.ts:327]: were you able to figure out why this happens?

File: packages/driver/src/cy/commands/sessions.ts [hunk 17]
```diff
@@ -0,0 +1,822 @@
...
+
+        const value = {} as any
+
+        if (_localStorage) {
+          value.localStorage = _localStorage
+        }
+
+        if (_sessionStorage) {
+          value.sessionStorage = _sessionStorage
+        }
```
Comment (flotwig) [inline packages/driver/src/cy/commands/sessions.ts:341]: so this is effectively "if there are any entries in local/sessionStorage, send local/sessionStorage", but working around the `Storage.length` thing you mentioned?

File: packages/driver/src/cy/commands/sessions.ts [hunk 18]
```diff
@@ -0,0 +1,822 @@
...
+        pushValue(current_origin, value)
+      }
+
+      if (_.isEmpty(origins)) {
+        return getResults()
+      }
+
+      if (current_origin.startsWith('https:')) {
+        _.remove(origins, (v) => v.startsWith('http:'))
+      }
```
Comment (flotwig) [inline packages/driver/src/cy/commands/sessions.ts:352]: any chance that silently excluding http origins from https origins could cause unexpected issues/confusion for users? this behavior doesn't seem to be documented in the docs PR

File: packages/driver/src/cy/commands/sessions.ts [hunk 19]
```diff
@@ -0,0 +1,822 @@
...
+
+    registerSessionHooks () {
+      Cypress.on('test:before:run:async', () => {
+        if (Cypress.config('experimentalSessionSupport')) {
+          currentTestRegisteredSessions.clear()
+
+          return navigateAboutBlank(false)
+          .then(() => sessions.clearCurrentSessionData())
+          .then(() => {
+            return Cypress.backend('reset:renderedHTMLOrigins')
```
Comment (flotwig) [inline packages/driver/src/cy/commands/sessions.ts:449]: love the mix of caterpillar-case and camel-case

```suggestion
            return Cypress.backend('reset:rendered:html:origins')
```

File: packages/driver/src/cy/commands/sessions.ts [hunk 20]
```diff
@@ -0,0 +1,822 @@
...
+        snapshot: false,
+      })
+
+      const dataLog = Cypress.log({
+        name: 'session',
+        sessionInfo: getSessionDetails(existingSession),
+        message: `${existingSession.id > 50 ? `${existingSession.id.substr(0, 47)}...` : existingSession.id}`,
+      })
+
+      async function runsetup (existingSession) {
```
Comment (flotwig) [inline packages/driver/src/cy/commands/sessions.ts:540]:
```suggestion
      async function runSetup (existingSession) {
```

File: packages/driver/src/cy/commands/sessions.ts [hunk 21]
```diff
@@ -0,0 +1,822 @@
...
+                indicator: 'successful',
+                message: `(new) ${_log.get().message}`,
+              }
+            },
+          })
+        }
+
+        await navigateAboutBlank()
+
+        cy.then(() => sessions.clearCurrentSessionData())
```
Comment (flotwig) [inline packages/driver/src/cy/commands/sessions.ts:563]: what's the reason to use `cy.then` instead of regular Promises/async/await inside of this function? at first i thought you were doing this because you used other chainables inside of cy.session, but i don't see any chainables besides `cy.then`, so it seems unnecessary to me to use it

File: packages/driver/src/cypress/error_messages.js [hunk 22]
```diff
@@ -101,7 +103,7 @@ const cyStripIndent = (str, indentSize) => {
   }).join('\n')
 }
 
-module.exports = {
+_.extend(module.exports, {
```
Comment (flotwig) [inline packages/driver/src/cypress/error_messages.js:106]: :thinking: why this change?

File: packages/driver/src/cypress/stack_utils.js [hunk 23]
```diff
@@ -357,19 +420,21 @@ const normalizedUserInvocationStack = (userInvocationStack) => {
   return normalizeStackIndentation(winnowedStackLines)
 }
 
-module.exports = {
+Object.assign(module.exports, {
```
Comment (flotwig) [inline packages/driver/src/cypress/stack_utils.js:423]: ditto

File: packages/proxy/lib/http/index.ts [hunk 24]
```diff
@@ -268,6 +275,8 @@ export class Http {
   }
 
   reset () {
+    // TODO: when to reset marked origins?
+    // this.renderedHTMLOrigins = {}
```
Comment (flotwig) [inline packages/proxy/lib/http/index.ts:279]: yes, when? this `Http` object's lifetime lasts for the entire time a project is open

File: packages/runner-shared/src/event-manager.js [hunk 25]
```diff
@@ -65,6 +65,8 @@ export const eventManager = {
       return this._reRun(state)
     }
 
+    top._rerun = rerun
```
Comment (flotwig) [inline packages/runner-shared/src/event-manager.js:68]: this appears to only be used in session_spec, does this need to be set in production?

File: packages/server/lib/browsers/cdp_automation.ts [hunk 26]
```diff
@@ -252,6 +277,21 @@ export class CdpAutomation {
 
           return this.sendDebuggerCommandFn('Network.deleteCookies', _.pick(cookieToBeCleared, 'name', 'domain'))
         })
+
+      case 'clear:cookies':
+        return Bluebird.mapSeries(data as CyCookieFilter[], async (cookie) => {
+          // resolve with the value of the removed cookie
+          // also, getting the cookie via CDP first will ensure that we send a cookie `domain` to CDP
+          // that matches the cookie domain that is really stored
```
Comment (flotwig) [inline packages/server/lib/browsers/cdp_automation.ts:285]: the amount of CDP requests we send for cookies stuff is crazy, sometimes the CDP debug logs scroll for minutes on the cookies e2e tests

File: packages/server/lib/routes.ts [hunk 27]
```diff
@@ -47,6 +47,74 @@ export const createRoutes = ({
...
+        value,
+        type: 'localStorage',
+      }, '*')
+    }).toString()})()</script></body></html>`)
+  })
+
+  /* eslint-disable no-undef */
+  app.get('/__cypress/automation/setLocalStorage', (req, res) => {
+    // const origin = req.host
+
```
Comment (flotwig) [inline packages/server/lib/routes.ts:77]:
```suggestion
```

File: packages/server/lib/routes.ts [hunk 28]
```diff
@@ -47,6 +47,74 @@ export const createRoutes = ({
...
+      const value = {} as any
+
+      if (_localStorage) {
+        value.localStorage = _localStorage
+      }
+
+      if (_sessionStorage) {
+        value.sessionStorage = _sessionStorage
+      }
```
Comment (flotwig) [inline packages/server/lib/routes.ts:65]: perhaps this would be good to deduplicate with the earlier usage of this pattern, or at least comment it, since at first glance it appears to be the work of a madperson

**Review by flotwig — CHANGES_REQUESTED** [review]
This is cool, looks like a good start. Mostly minor things.

Now, after reviewing 4080 lines and 78 files changed, I am gonna go take a break :smile:

File: packages/server/lib/routes.ts (hunk 28 above)
Comment (Bkucera) [reply packages/server/lib/routes.ts:67]: we cant really dedupe here since this function is stringified and evalled in cross origin iframes. added comment

File: packages/runner-shared/src/event-manager.js (hunk 25 above)
Comment (Bkucera) [reply packages/runner-shared/src/event-manager.js]: 85d85eaeeb

File: packages/server/lib/routes.ts (hunk 27 above)
Comment (Bkucera) [reply packages/server/lib/routes.ts]: 85d85eaeeb

File: packages/driver/src/cy/commands/sessions.ts (hunk 14 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts:178]: 85d85eaeeb

File: packages/driver/src/cy/commands/sessions.ts (hunk 8 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts]: 85d85eaeeb

File: packages/driver/src/cy/commands/sessions.ts (hunk 9 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts:175]: 85d85eaeeb

File: packages/driver/src/cy/commands/local_storage.js (hunk 10 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/local_storage.js]: 85d85eaeeb

File: packages/server/test/support/fixtures/projects/session/cypress/integration/spec.js (hunk 12 above)
Comment (Bkucera) [reply packages/server/test/support/fixtures/projects/session/cypress/integration/spec.js]: 85d85eaeeb

File: cli/types/cypress.d.ts [hunk 29]
```diff
@@ -952,6 +970,13 @@ declare namespace Cypress {
      */
     debug(options?: Partial<Loggable>): Chainable<Subject>
 
+    /**
+     * Apply a Session's cookie/localStorage data to the current test
+     *
+     * @see https://on.cypress.io/session
+     */
+    session(id: string, setup?: SessionOptions['validate'], options?: SessionOptions): Chainable<null>
+
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply cli/types/cypress.d.ts]: 85d85eaeeb

File: packages/driver/cypress/integration/e2e/e2e_cookies_spec.js [hunk 30]
```diff
@@ -82,5 +94,27 @@ describe('e2e cookies spec', () => {
         done()
       })
     })
+
+    // a hostOnly cookie is set in the browser by omitting the 'domain' property
+    // so for cross-domain cookies users need to add hostOnly: true as well as 'domain'
+    // TODO: un-skip when we rework cookie API to support setting hostOnly thirdparty cookies.
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply packages/driver/cypress/integration/e2e/e2e_cookies_spec.js:100]: 85d85eaeeb

File: packages/driver/src/cy/commands/cookies.js [hunk 31]
```diff
@@ -65,6 +65,20 @@ function cookieValidatesSecurePrefix (options) {
   return options.secure === false
 }
 
+// TODO: do something with the hostOnly property. Right now we dont return it
+// to prevent changes to the API. We will rework cookies in the future.
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply packages/driver/src/cy/commands/cookies.js]: 85d85eaeeb

File: packages/driver/src/cy/commands/sessions.ts (hunk 13 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts]: 85d85eaeeb

File: packages/driver/src/cy/commands/sessions.ts (hunk 6 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts]: 85d85eaeeb

File: packages/driver/src/cy/commands/sessions.ts (hunk 5 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts]: 85d85eaeeb

File: packages/driver/src/cy/commands/sessions.ts (hunk 4 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts]: 85d85eaeeb

File: packages/driver/src/cy/commands/sessions.ts [hunk 32]
```diff
@@ -0,0 +1,822 @@
...
+    },
+  })
+
+  Cypress.session = sessions
+}
+
+function navigateAboutBlank (session = true) {
+  // if (cy.state('window').location.href === 'about:blank') {
+  //   return Promise.resolve()
+  // }
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts]: 85d85eaeeb

File: packages/proxy/lib/http/index.ts (hunk 24 above)
Comment (Bkucera) [reply packages/proxy/lib/http/index.ts:279]: 461e15c28

Comment (Bkucera) [reply packages/proxy/lib/http/index.ts:279]: 461e15c28

File: packages/driver/src/cy/commands/sessions.ts (hunk 7 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts]: 64485d1d updated comment

**Review by chrisbreiding — DISMISSED** [review]
I think there are some items from Zach's review that haven't been addressed (I think GitHub collapsed them, so maybe they were missed), but it looks like everything from my review was addressed.

File: packages/driver/src/cy/commands/sessions.ts (hunk 15 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts:382]: addressed

File: packages/driver/src/cy/commands/sessions.ts (hunk 16 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts]: no

File: packages/driver/src/cy/commands/sessions.ts (hunk 17 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts]: yes

File: packages/driver/src/cy/commands/sessions.ts (hunk 19 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts]: fixed

File: packages/driver/src/cy/commands/sessions.ts (hunk 21 above)
Comment (Bkucera) [reply packages/driver/src/cy/commands/sessions.ts]: fixed

File: packages/driver/src/cypress/error_messages.js (hunk 22 above)
Comment (Bkucera) [reply packages/driver/src/cypress/error_messages.js:109]: added comment

File: packages/driver/src/cypress/stack_utils.js (hunk 23 above)
Comment (Bkucera) [reply packages/driver/src/cypress/stack_utils.js]: added comment

File: packages/driver/src/cy/commands/sessions.ts (hunk 16 above)
Comment (flotwig) [reply packages/driver/src/cy/commands/sessions.ts]: fml lol

**Review by flotwig — DISMISSED** [review]
noice

**Review by chrisbreiding — DISMISSED** [review]

**Review by flotwig — APPROVED** [review]
